application/src/main/java/org/opentripplanner/graph_builder/module/osm/OsmModule.java

```diff
@@ -545,6 +569,95 @@ private Optional<Platform> getPlatform(OsmDatabase osmdb, OsmWay way) {
+    List<OsmLevel> levels = osmdb.getLevelsForEntity(way);
+    var nodeRefs = way.getNodeRefs();
+    if (nodeRefs.size() > 1) {
+      long firstNodeRef = nodeRefs.get(0);
+      long lastNodeRef = nodeRefs.get(nodeRefs.size() - 1);
+      if (levels.size() == 2) {
+        OsmLevel firstVertexOsmLevel = levels.get(0);
+        OsmLevel lastVertexOsmLevel = levels.get(1);
+        if (firstVertexOsmLevel.level() < lastVertexOsmLevel.level()) {
+          return Optional.of(
+            new EdgeLevelInfo(
+              new VertexLevelInfo(
+                firstVertexOsmLevel.level(),
+                firstVertexOsmLevel.name(),
+                firstNodeRef
+              ),
+              new VertexLevelInfo(
+                lastVertexOsmLevel.level(),
+                lastVertexOsmLevel.name(),
+                lastNodeRef
+              )
+            )
+          );
+        } else if (firstVertexOsmLevel.level() > lastVertexOsmLevel.level()) {
+          return Optional.of(
+            new EdgeLevelInfo(
+              new VertexLevelInfo(
+                lastVertexOsmLevel.level(),
+                lastVertexOsmLevel.name(),
+                lastNodeRef
+              ),
+              new VertexLevelInfo(
+                firstVertexOsmLevel.level(),
+                firstVertexOsmLevel.name(),
+                firstNodeRef
+              )
+            )
+          );
+        }
+      }
+      if (way.hasTag("incline")) {
+        if (way.isInclineUp()) {
+          return Optional.of(
+            new EdgeLevelInfo(
+              new VertexLevelInfo(null, null, firstNodeRef),
+              new VertexLevelInfo(null, null, lastNodeRef)
+            )
+          );
+        } else if (way.isInclineDown()) {
+          return Optional.of(
+            new EdgeLevelInfo(
+              new VertexLevelInfo(null, null, lastNodeRef),
+              new VertexLevelInfo(null, null, firstNodeRef)
+            )
+          );
+        }
+      }
+    }
+    return Optional.empty();
+  }
+
+  private void decorateEdgeWithLevelInfo(
+    @Nullable Edge forwardEdge,
+    @Nullable Edge backwardEdge,
+    Optional<EdgeLevelInfo> edgeLevelInfoOptional,
+    OsmWay way
+  ) {
+    if (params.includeEdgeLevelInfo() && edgeLevelInfoOptional.isPresent()) {
+      EdgeLevelInfo edgeLevelInfo = edgeLevelInfoOptional.get();
+      Edge edge = forwardEdge != null ? forwardEdge : backwardEdge;
+      if (
+        edge != null &&
+        edge.getToVertex() instanceof OsmVertex toVertex &&
+        edge.getFromVertex() instanceof OsmVertex fromVertex &&
+        edgeLevelInfo.matchesNodes(fromVertex.nodeId, toVertex.nodeId)
+      ) {
+        if (forwardEdge != null) {
+          osmStreetDecoratorRepository.addEdgeLevelInformation(forwardEdge, edgeLevelInfo);
+        }
+        if (backwardEdge != null) {
+          osmStreetDecoratorRepository.addEdgeLevelInformation(backwardEdge, edgeLevelInfo);
+        }
```

Comment:
This block reads a bit weird as we check that one of the edges matches the criteria and then we do some stuff for both. However, I couldn't come up with a solution that would be clearly more readable if we actually need the weird checks that I've mentioned above.

Reply:
I agree that it does look interesting